twine/settings.py

```diff
@@ -338,6 +341,23 @@ def check_repository_url(self) -> None:
                 repository_url, utils.DEFAULT_REPOSITORY, utils.TEST_REPOSITORY
             )
 
+        repository_host = urlparse(repository_url).hostname
+        if (
+            repository_host
+            and repository_host.endswith("pypi.org")
+            and repository_url not in [utils.DEFAULT_REPOSITORY, utils.TEST_REPOSITORY]
+        ):
+            raise exceptions.InvalidPyPIUploadURL(
+                f"The configured repository {repository_url} is not a known "
+                f"PyPI repository.\n"
+                f"You probably want {utils.DEFAULT_REPOSITORY} "
+                f"or {utils.TEST_REPOSITORY}.\n"
+                f"Check your --repository-url value "
+                f"or the repository configuration in {self.config_file}.\n"
+                f"See https://packaging.python.org/specifications/pypirc/ "
+                f"for more details."
+            )
+
```

Comment:
I fixed the failing test and adding coverage for the new behavior in https://github.com/pypa/twine/pull/809/commits/414625b1414efbc8c939a2e212fc6ba753dde0f0, but looking at the drop in coverage [via Codecov](https://app.codecov.io/gh/pypa/twine/compare/809/changes#D1L189) emphasized my point in https://github.com/pypa/twine/pull/809#discussion_r706853387, i.e. that this is similar to [`utils.check_status_code`](https://github.com/pypa/twine/blob/f54311486f1a27e732fa0118ac1b2b71f1c32c10/twine/utils.py#L172).  

Taking a closer look at that, plus the comments in https://github.com/pypa/twine/issues/310, plus my initial attempt in https://github.com/pypa/twine/pull/460 to offer a better error message, I'm now wondering if it's needlessly redundant, with the primary difference being that this executes _before_ an upload, while `check_status_code` executes afterwards. Instead, maybe we could improve the post-upload `RedirectDetected` message by tweaking the language, possibly with a special message for a trailing slash, regardless of whether or not it's a `pypi.org` hostname.

In other words, this feels like a case of "Look Before You Leap" vs "It's Easter to Ask Forgiveness than Permission", and with this PR, I think we're doing both. I'm going to mull this over a bit, and maybe put up another PR that improves the EAFP approach.

@meowmeowmeowcat Apologies for the confusion here; I should have taken a closer look at the existing behavior.

Reply:
@bhrutledge I think the improved message in #812 would be also helpful:
```
RedirectDetected: https://upload.pypi.org/legacy attempted to redirect to https://upload.pypi.org/legacy/.
Your repository URL is missing a trailing slash. Please add it and try again.
```

But I also think that having this check will save folks' time. They will know there is an error in the config or `--repository-url`` before twine tries to upload their distributions.

Reply:
I see. Probably this PR can be closed. The improved message in #812 is enough.

Reply:
I'd like to get @sigmavirus24 feedback on #812 first.